kafkaesk/retry.py

```diff
@@ -0,0 +1,295 @@
+                    handler = self._handlers[handler_exception_type]
+                    handler_key = handler_exception_type.__name__
+                    self._handler_cache[exception_type] = (handler_key, handler)
+                    break
+
+        return (handler_key, handler)
+
+
+class RetryHandler(ABC):
+    """Base class implementing common logic for RetryHandlers
+
+    All RetryHandler's should implement the following metrics:
+
+    * RETRY_HANDLER_FORWARD - Note: This is implemented by RetryHandler._forward_message
+    * RETRY_HANDLER_DROP - Note: this is implemented by RetryHandler._drop_message
+    * RETRY_CONSUMER_TOPIC_OFFSET
+    * RETRY_CONSUMER_MESSAGE_LEAD_TIME - Note: If a RetryHandler's consumer expects a delay,
+        this delay should be subtracted from the lead time
+    * RETRY_CONSUMER_CONSUMED_MESSAGE_TIME
+    * RETRY_CONSUMER_CONSUMED_MESSAGES
+
+    See `kafkaesk.metrics` for more information on each of these metrics
+    """
+
+    def __init__(self, stream_id: str) -> None:
+        self.stream_id = stream_id
+
+        self._ready = False
+
+    async def initialize(self, policy: RetryPolicy) -> None:
+        self._ready = True
+
+    async def finalize(self) -> None:
+        self._ready = False
+
+    async def __call__(
+        self,
+        policy: RetryPolicy,
+        handler_key: str,
+        retry_history: RetryHistory,
+        record: ConsumerRecord,
+        exception: Exception,
+    ) -> None:
+        raise NotImplementedError
+
+    async def _drop_message(
+        self,
+        policy: RetryPolicy,
+        retry_history: RetryHistory,
+        record: ConsumerRecord,
+        exception: Exception,
+    ) -> None:
+        logger.info(
+            f"{self.__class__.__name__} handler recieved exception, dropping message",
+            exc_info=exception,
+        )
+        RETRY_HANDLER_DROP.labels(
+            stream_id=record.topic,
+            partition=record.partition,
+            group_id=policy.subscription.group,
+            handler=self.__class__.__name__,
+            exception=exception.__class__.__name__,
+        ).inc()
+
+    async def _forward_message(
+        self,
+        policy: RetryPolicy,
+        retry_history: RetryHistory,
+        record: ConsumerRecord,
+        exception: Exception,
+        forward_stream_id: str,
+    ) -> None:
+        await policy.app.publish(
+            forward_stream_id,
+            RetryMessage(
+                original_record=Record.from_consumer_record(record), retry_history=retry_history
+            ),
+        )
+        RETRY_HANDLER_FORWARD.labels(
+            stream_id=record.topic,
+            partition=record.partition,
+            group_id=policy.subscription.group,
+            handler=self.__class__.__name__,
+            exception=exception.__class__.__name__,
+            forward_stream_id=forward_stream_id,
+        ).inc()
+
+
+class NoRetry(RetryHandler):
```

Comment:
Can we rename this to `Drop`? `NoRetry` to me might sound like it would mean the exception would bubble.

Reply:
Good call, done.